test/SILOptimizer/utf8_decoding_fastpath.swift

```diff
@@ -0,0 +1,79 @@
+// good code for various ways to create a String from a Sequence of UTF-8
+// bytes for which a fast path exists.
+
+// Please note: this test targets "core2" to ensure consistent output
+// on all x86 host processors.
+
+// UnsafeBufferPointer<UInt8>: we check that we call the merged function
+// ==========================
+// CHECK-LABEL: define swiftcc {{.*}}s22utf8_decoding_fastpath15decodeUBPAsUTF8ySSSRys5UInt8VGF{{.*}}
+// CHECK-NEXT: {{.*}} = tail call swiftcc {{.*}}s22utf8_decoding_fastpath15decodeUBPAsUTF8ySSSRys5UInt8VGFTm{{.*}}
+// CHECK: ret
+public func decodeUBPAsUTF8(_ ptr: UnsafeBufferPointer<UInt8>) -> String {
+  return String(decoding: ptr, as: Unicode.UTF8.self)
+}
+
+// UnsafeMutableBufferPointer<UInt8>: we check that we call the merged function
+// =================================
+// CHECK-LABEL: define swiftcc {{.*}}s22utf8_decoding_fastpath16decodeUMBPAsUTF8ySSSrys5UInt8VGF{{.*}}
+// CHECK-NEXT: {{.*}} = tail call swiftcc {{.*}}s22utf8_decoding_fastpath15decodeUBPAsUTF8ySSSRys5UInt8VGFTm{{.*}}
+// CHECK: ret
+public func decodeUMBPAsUTF8(_ ptr: UnsafeMutableBufferPointer<UInt8>) -> String {
+  return String(decoding: ptr, as: Unicode.UTF8.self)
+}
+
+// this defined the merged function that handles both UBP and UMBP
+// CHECK-LABEL: define internal swiftcc {{.*}}s22utf8_decoding_fastpath15decodeUBPAsUTF8ySSSRys5UInt8VGFTm{{.*}}
+// CHECK-NOT: {{.*}} = call swiftcc {{.*}}_fromCodeUnits{{.*}}
+// CHECK: {{.*}} = call swiftcc {{.*}} @"$sSS18_fromUTF8Repairing{{.*}}
+// CHECK-NOT: {{.*}} = call swiftcc {{.*}}_fromCodeUnits{{.*}}
+// CHECK: ret
+
+// Array<UInt8>
+// ============
+// CHECK-LABEL: define swiftcc {{.*}}decodeArrayAsUTF8{{.*}}
+// CHECK-NOT: {{.*}} = call swiftcc {{.*}}_fromCodeUnits{{.*}}
+// CHECK: {{.*}} = call swiftcc {{.*}} @"$sSS18_fromUTF8Repairing{{.*}}
+// CHECK-NOT: {{.*}} = call swiftcc {{.*}}_fromCodeUnits{{.*}}
+// CHECK: ret
+public func decodeArrayAsUTF8(_ ptr: [UInt8]) -> String {
+  return String(decoding: ptr, as: Unicode.UTF8.self)
+}
+
+// UnsafeRawBufferPointer
+// ======================
+// CHECK-LABEL: define swiftcc {{.*}}decodeURBPAsUTF8{{.*}}
+// CHECK-NOT: {{.*}} = call swiftcc {{.*}}_fromCodeUnits{{.*}}
+// CHECK: {{.*}} = call swiftcc {{.*}} @"$sSS18_fromUTF8Repairing{{.*}}
+// CHECK-NOT: {{.*}} = call swiftcc {{.*}}_fromCodeUnits{{.*}}
+// CHECK: ret
+public func decodeURBPAsUTF8(_ ptr: UnsafeRawBufferPointer) -> String {
+  return String(decoding: ptr, as: Unicode.UTF8.self)
+}
+
+// UnsafeMutableRawBufferPointer
+// =============================
+// CHECK-LABEL: define swiftcc {{.*}}decodeUMRBPAsUTF8{{.*}}
+// CHECK-NOT: {{.*}} = call swiftcc {{.*}}_fromCodeUnits{{.*}}
+// CHECK: {{.*}} = call swiftcc {{.*}} @"$sSS18_fromUTF8Repairing{{.*}}
+// CHECK-NOT: {{.*}} = call swiftcc {{.*}}_fromCodeUnits{{.*}}
+// CHECK: ret
+public func decodeUMRBPAsUTF8(_ ptr: UnsafeMutableRawBufferPointer) -> String {
+  return String(decoding: ptr, as: Unicode.UTF8.self)
+}
+
```

Comment:
What did String and Substring end up looking like? Would it be easy to check contingent on the `isFastUTF8` bit?

Reply:
@milseman what do you mean by `String` and `SubString` here? They're not sequences of `UInt8` right?

Reply:
I mean, `String.utf8` and `Substring.utf8`?

Reply:
@milseman added `String.UTF8View` and `Substring.UTF8View` (also don't work)

Reply:
done